docs:#602 basic auth for proxy and schema registry (#718)
redpanda-data/documentation: docs/security/authentication.mdx

<head>
    <meta name="title" content="Configuring Authentication | Redpanda Docs"/>

>     <meta name="description" content="Redpanda supports multiple forms of authentication including SASL/SCRAM, basic authentication, mTLS with principal mapping, and mixed-mode with multiple listeners."/>

Reviewer: This mixes a few things (not all endpoints support all forms of auth).

Maybe we should say something about 'Different components of Redpanda support a wide variety of authentication methods...'

Maybe it's too detailed but it may be worth specifying that 'listeners' relate specifically to  the Kafka API endpoint.

Reviewer 2: >  and mixed-mode with multiple listeners.

This is a bit awkward, mixed-mode isn't really a thing that's like the others; each listener has an auth mode and there can be multiple listeners.

> Maybe it's too detailed but it may be worth specifying that 'listeners' relate specifically to the Kafka API endpoint.

Multiple listeners is supported for Kafka API, Schema Registry, and HTTP Proxy (at least in theory, I don't know if helm and operator directly support that).

Author: @mattschumpert or @BenPope I've updated the intro paragraph: https://deploy-preview-718--redpanda-documentation.netlify.app/docs/security/authentication/. Is this correct?

### Create superusers

When you configure authentication, you include one or more superusers in the Redpanda configuration file. This user has [ALL permissions](../../security/authorization#operations) on the cluster and grants permissions to new users. 
(Without a superuser, you can create other users, but you can't grant them permissions to the cluster.)

> Specify the name of the superuser, either an existing user or a user that does not exist yet. For example, if you used superuser `admin`, Redpanda allows the `admin` user to do anything, but Redpanda does not create the `admin` user. To create this superuser, run: 

Reviewer 3: I think it would be nice/helpful to our customers if there were more background/context here that answers the questions:
- Why should our customers create/include superusers in the RP config file?
 Here, the parenthetical kind of touches on it--is the parenthetical the primary reason our customers need to create superusers? If so, maybe present as the explanation? eg. "When you configure authentication, it is highly recommended that you include one or more superusers in the Redpanda configuration file. This user has [ALL permissions](../../security/authorization#operations) on the cluster and grants permissions to new users. 
Without a superuser, you can create other users, but you can't grant them permissions to the cluster.

Author: rephrased this

```bash
rpk acl user create <superuser_username> -p <superuser_password> 
```

> The `admin` user has full access to the cluster and can then grant permissions to other users. 

Reviewer 3: ```suggestion
The `admin` user has full access to the cluster and can grant permissions to other users. 
```

Author: fixed

For information about using `rpk` to manage ACL users, see [rpk acl](../../reference/rpk/rpk-acl/). 

> As a security best practice, you don't want to use the superuser to interact with the cluster, but you also don't want to delete the superuser (in case you need to grant permissions to new users later on). In addition, when you create the superuser, you specify a password, which adds a level of security to the superuser. If you delete the user, someone else might re-create the user with a different password. 

Reviewer 3: ```suggestion
As a security best practice, don't use the superuser to interact with the cluster and do not delete the superuser (in case you need to grant permissions to new users later on). In addition, when you create the superuser, specify a password, which adds a level of security to the superuser. If you delete the user, someone else might re-create the user with a different password. 
```

Author: fixed!

// highlight-end
```

> ### Configure SASL for Schema Registry and HTTP Proxy

Reviewer 2: ```suggestion
### Configure Schema Registry and HTTP Proxy to connect to Redpanda with SASL
```

> Schema Registry and HTTP Proxy connect to Redpanda over the Kafka API. For the Kafka username and password, Redpanda uses ephemeral secrets internal to the cluster, which change on cluster restart. When a node receives a Kafka request with a username that looks internal, it consults the in-memory store of ephemeral secrets to authenticate. If there is no secret matching the name, it broadcasts to other nodes to share the secret. If no secret is available, it asks the controller leader to mint a new secret.

Reviewer: This page is good place to document it @Feediver1 .

 ephemeral credentials ('AutoAuth') is still landing this week, but we should probably have a separate PR for this as it' arguably its own 'feature' @micheleRP .  But I still think the 'authentication' page is the right context in which to read about it

> Schema Registry and HTTP Proxy connect to Redpanda over the Kafka API. For the Kafka username and password, Redpanda uses ephemeral secrets internal to the cluster, which change on cluster restart. When a node receives a Kafka request with a username that looks internal, it consults the in-memory store of ephemeral secrets to authenticate. If there is no secret matching the name, it broadcasts to other nodes to share the secret. If no secret is available, it asks the controller leader to mint a new secret.

Author: @Feediver1 @mattschumpert We already have https://github.com/redpanda-data/documentation/issues/607, which I included in this PR with basic auth.

### Configure SASL for HTTP Proxy and Schema Registry

> HTTP Proxy and Schema Registry connect to Redpanda over the Kafka API, and they must be configured with a username and password.

Reviewer: @micheleRP this doesn't talk about ephemeral credentials (the other option here that makes this statement not true).  I thought we were adding that content in this PR?

> For Schema Registry, the configuration node is `schema_registry_client`:

Reviewer: @BenPope  a nit, it's a bit confusing that we're talking about SR/PP itself and not a client TO them with this conf.

Reviewer 2: Yeah, this could probably be clearer.

`Configure HTTP Proxy and Schema Registry to connect to Redpanda with SASL`

Author: thx, fixed

// highlight-end
```

### Configure Schema Registry and HTTP Proxy to connect to Redpanda with SASL

> Schema Registry and HTTP Proxy connect to Redpanda over the Kafka API. For the Kafka username and password, Redpanda uses ephemeral credentials internal to the cluster. Ephemeral credentials are regular SCRAM credentials, but they're only stored in memory and are lost when a node restarts. When the Schema Registry or HTTP Proxy start up, they broadcast an ephemeral credential to other nodes over the internal RPC. If authentication fails to a particular broker, new ephemeral credentials are sent to that broker, and the service reconnects.

Reviewer 3: So much better/more clear/helpful!

message.timestamp.type  CreateTime  DEFAULT_CONFIG
...
```

## Configure basic authentication

:::note

> Basic authentication is supported on the Admin API, HTTP Proxy, and Schema Registry.

Reviewer: This section doesn't address how basic auth on the Admin API is configured.  Is it just 'always-on' @BenPope ?  If so, we should say that here for completeness @micheleRP

Reviewer 2: [admin_api_require_auth](https://docs.redpanda.com/docs/reference/cluster-properties/#:~:text=admin_api_require_auth)

Author: > [admin_api_require_auth](https://docs.redpanda.com/docs/reference/cluster-properties/#:~:text=admin_api_require_auth)

@BenPope Would you please confirm if this is now correct: https://deploy-preview-718--redpanda-documentation.netlify.app/docs/security/authentication/#configure-basic-authentication

message.timestamp.type  CreateTime  DEFAULT_CONFIG
...
```

## Configure basic authentication

:::note
Basic authentication is supported on the Admin API, Schema Registry, and HTTP Proxy.
:::

To configure basic authentication on the Admin API, set [`admin_api_require_auth`](../../reference/cluster-properties/#admin) to `true`. 
Administrators create users with [`rpk acl user create`](../../reference/rpk/rpk-acl/rpk-acl-user-create/). This adds users to the Redpanda credential store that HTTP basic authentication uses.

> You can enable basic authentication to use Kafka API username/password credentials to authenticate to HTTP Proxy and Schema Registry. 

Reviewer 4: ```suggestion
You can enable basic authentication to use Kafka API username/password credentials to authenticate to HTTP Proxy and Schema Registry. This requires that SASL is turned on for Kafka API endpoints.
```

Both the HTTP Proxy and Schema Registry require that SASL is enabled on Kafka API endpoints

> - HTTP Proxy: Users represented in the HTTP basic authorization header are subject to authorization restrictions by Redpanda ACLs. Access to the Kafka API impersonates the user whose credentials were used to authenticate to HTTP Proxy. Redpanda passes the username/password in memory to a SASL-enabled Kafka client. This requires that SASL is turned on for Kafka API endpoints.

Reviewer 4: It seems that this bullet point is intended to describe **how** we implemented basic authentication on the HTTP Proxy. With that in mind, I suggest a slight re-ordering to the sentences so readers can differ between the conceptual and implementation elements. We also don't need to mention the authorization header here since it is an industry standard.

```suggestion
- HTTP Proxy: Access to the Kafka API impersonates the user whose credentials were used to authenticate to HTTP Proxy and the user is subject to authorization restrictions by Redpanda ACLs. To support this design, Redpanda passes the username/password in memory to a SASL-enabled Kafka client.
```

- Schema Registry: Authorization is "all or nothing": if the user presents a valid user account, they have full read/write access.

To configure basic authentication, set `authentication_method` to `http_basic`.

In `redpanda.yaml`, enter:
```yaml
pandaproxy:
  pandaproxy_api:
  - address: "localhost"
    port: 8082
    authentication_method: http_basic

schema_registry:
  schema_registry_api:
    address: "localhost"
    port: 8081
    authentication_method: http_basic
```    

> 

Reviewer 4: Also, here is an example of how to use basic authentication after the feature is enabled. This could help with understanding.

```
rpk acl user create foo --password bar # Creates a SASL user for the Kafka API
curl -u "foo:bar" "http://localhost:8082/topics" # A request to the HTTP Proxy with user "foo"
curl -u "foo:bar" "http://localhost:8081/subjects" # A request to the Schema Registry with user "foo" 
```

Author: Thanks @NyaliaLui! Can you please confirm if this looks correct, just dropping it in? https://deploy-preview-718--redpanda-documentation.netlify.app/docs/security/authentication/#configure-basic-authentication

Reviewer 4: Looks good.

```
See also: [Configuring Redpanda SASL on Kubernetes](../../reference/redpanda-operator/kubernetes-sasl)

> ## Configure basic authentication

Reviewer 4: It may be worth noting that Basic Authentication is only supported on the Admin API, HTTP Proxy, and Schema Registry.

You can enable basic authentication to use Kafka username/password credentials to authenticate to HTTP Proxy and Schema Registry. Administrators create users with `rpk acl user create`. This adds users to the Redpanda credential store that HTTP basic authentication uses.

- HTTP Proxy: Users represented in the basic authorization header are subject to authorization restrictions by Redpanda ACLs. Access to the Kafka API impersonates the user whose credentials were used to authenticate to HTTP Proxy. Redpanda passes the username/password in memory to a SASL-enabled Kafka API endpoint. This requires SASL enabled on Kafka API endpoints.

- Schema Registry: Authorization is "all or nothing": if the user presents a valid user account, they have full read/write access.

> To configure basic authentication, set `authentication_type` of the listener to `http_basic`.

Reviewer 4: ```suggestion
To configure basic authentication, set `authentication_method` of the listener to `http_basic`.
```

### Configure mTLS for HTTP Proxy and Schema Registry

> The HTTP Proxy and Schema Registry require valid client certificates.

Reviewer 2: The example connects Schema Registry to Redpanda with mTLS. If the `kafka_api` listener it connects to has `mtls_identity`, then that will be used. I wrote this part of the doc to highlight the second scenario.

> The HTTP Proxy and Schema Registry require valid client certificates.

Reviewer 2: Looks good

rpk cluster config set kafka_mtls_principal_mapping_rules '["DEFAULT"]'
```

### Configure mTLS for Schema Registry and HTTP Proxy

> Schema Registry and HTTP Proxy require valid client certificates to secure the transport. Continuing with the previous example, where the certificate contains an identity for authentication (`kafka_api` listener set to `mtls_identity`), the following example connects Schema Registry and HTTP Proxy to Redpanda with mTLS certificate-based identity.

Reviewer 3: ```suggestion
Schema Registry and HTTP Proxy require valid client certificates to secure the transport. Continuing with the previous example, where the certificate contains an identity for authentication (`kafka_api` listener set to `mtls_identity`), the following example shows how to connect Schema Registry and HTTP Proxy to Redpanda with mTLS certificate-based identity.
```

rpk cluster config set kafka_mtls_principal_mapping_rules '["DEFAULT"]'
```

> ### Configure mTLS for Schema Registry and HTTP Proxy

Reviewer 2: ```suggestion
### Configure Schema Registry and HTTP Proxy to connect to Redpanda with mTLS
```

> Schema Registry and HTTP Proxy require valid client certificates to secure the transport. Continuing with the previous example, where the certificate contains an identity for authentication (`kafka_api` listener set to `mtls_identity`), the following example shows how to connect Schema Registry and HTTP Proxy to Redpanda with mTLS certificate-based identity.

Reviewer 2: Looks good.

Perhaps: "Schema Registry and HTTP Proxy require valid client certificates to secure the connection to Redpanda."

// highlight-end
```

## Disable authentication

> To disable authentication for a listener, set `authentication_method` to `none`. If authorization is enabled, connections to this listener use the anonymous user.

Reviewer: Maybe an example?

Reviewer 4: To make it easier, here is what the example would look like:
```yaml
pandaproxy:
  pandaproxy_api:
  - address: "localhost"
    port: 8082
    authentication_method: none

schema_registry:
  schema_registry_api:
    address: "localhost"
    port: 8081
    authentication_method: none
```

Author: @NyaliaLui thank you!

// highlight-end
```

## Disable authentication

> To disable authentication for a listener, set `authentication_method` to `none`:

Reviewer 4: Disabling authentication on the Kafka API is slightly more involved than on the Schema Registry and HTTP Proxy. It may be worth noting that here.

For the Kafka API, the user disables authentication by disabling `kafka_enable_authorization` and setting `authentication_method` to none for all listeners. For example:
`rpk cluster config set kafka_enable_authorization false` and
```yaml
redpanda:
  kafka_api:
    - address: 0.0.0.0
      port: 9092
      name: sasl_listener
      authentication_method: none
```

The writing and examples for disabling authentication on the Proxy and Schema Registry are correct so no changes there.

Author: thanks @NyaliaLui. I added this:  https://deploy-preview-718--redpanda-documentation.netlify.app/docs/security/authentication/#disable-authentication

Reviewer 4: Looks good.
